scio-bigtable/src/main/java/com/spotify/scio/bigtable/BigtableDoFn.java

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright 2017 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.bigtable;
+
+import com.google.cloud.bigtable.config.BigtableOptions;
+import com.google.cloud.bigtable.grpc.BigtableSession;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Queues;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.KV;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Semaphore;
+import java.util.function.Consumer;
+
+/**
+ * A {@link DoFn} that performs asynchronous lookup using Google Cloud Bigtable.
+ * @param <A> input element type.
+ * @param <B> Bigtable lookup value type.
+ */
+public abstract class BigtableDoFn<A, B> extends DoFn<A, KV<A, BigtableDoFn.Try<B>>> {
+  private static final Logger LOG = LoggerFactory.getLogger(BigtableDoFn.class);
+
+  // DoFn is deserialized once per CPU core. We assign a unique UUID to each DoFn instance upon
+  // creation, so that all cloned instances share the same ID. This ensures all cores share the
+  // same BigtableSession and Cache.
+  private static final ConcurrentMap<UUID, BigtableSession> session = Maps.newConcurrentMap();
+  private static final ConcurrentMap<UUID, Cache> cache = Maps.newConcurrentMap();
+  private final UUID instanceId;
+
+  private final BigtableOptions options;
+  private final CacheSupplier<A, B, ?> cacheSupplier;
+
+  // Data structures for handling async requests
+  private final Semaphore semaphore;
```

Comment:
@nevillelyh since we are exposing `BigtableSession` users have access to throttled client. Do you think it's better to have this extra guard?

Reply:
Actually prefer to leave this here. `BigtableSession` has a lot of methods, including 2 data clients and is confusing. This gives us extra guarantee that the `DoFn` won't blow up and be killed by the runner.